Required and Science command

yan-soon opened the pull request:
**RequiredCommand**

- Added Required Command that showcases all the leftover modules to take after cross referencing against user's GradPad
- Can be accessed by typing 'required'
- Modules are split into 5 various categories
    1. Foundation
    2. IT Professionalism
    3. Math And Science
    4. Science
    5. Industrial Experience
- Respective modules can be found under data folder
- If user has completed all modules, a success message will be shown for each category(eg. "You have completed all your Foundation Modules!")
- Focus Area modules have not been added (KIV)
- For Industrial Experience, 12MCs are needed, thus, the required command also calculates the current MC the user is at at reflects it in the Command Line Display

**ScienceCommand**

-Added Science command that displays all the available Science modules.
-Can be accessed by typing 'science'
- Respective modules can be found under data folder as well

**General Design Approach**

-The RequiredCommand and ScienceCommand have various attributes to help store the modules, namely a ReadOnlyGradPad for storage of modules and an actual ObservableList<Module> to do the comparisons with.
-Each module category will have a method to set up the module storage(eg. setFoundationStorage) and compare that storage against the user's currents modules (compareFoundation) to see if the user is missing anything out.
-The execute method then invokes all the various methods to set up the storages and does the comparisons, keeping track of all the undone modules, before displaying it to the user.


**Notes**

Since the command works by retrieving the stored modules in a JSON file and that the RequiredCommand extends the Command class, I had to change the execute method, to allow it to throw IOExceptions and DataConversionExceptions so that it can handle exceptions if the JSON file is not found. As a result, I changed a couple methods to throw IOExceptions and DataConversionExceptions, but this should not affect the general flow of the app.

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+            + "Math and Science Module Requirement!";
+    public static final String MESSAGE_FAILURE_ITPROF = "There was an error loading the IT Professionalism Modules :(";
+    public static final String MESSAGE_FAILURE_MATHANDSCI = "There was an error loading"
+            + " the Math and Science Modules :(";
```
Since we have so many messages here, perhaps we can shift this into a class on its own e.g. `RequiredCommandMessages` to avoid bloating the `RequiredCommand` class. We can then do static imports as needed

yan-soon replied:
Alright, made a new class `RequiredCommandMessages` in the storage component!

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+                String moduleToAdd = module.getModuleCode() + " (" + module.getModularCredits() + " MCs)";
+                modulesToAdd += "\n" + moduleToAdd;
+                foundationClear = false;
+            }
```
This code seems to be repeated in a few `compareX` fxn so it might be good to abstract it out into a `getModulesToAdd(<requiredModuleList>)` fxn or something along those lines that takes in a List and returns the string `modulesToAdd`

yan-soon replied:
True! I combined the 3 methods into one (those that compare foundation, ITprof and mathandsci mods) into one method. I decided to keep the compareInternship() and compareScience() methods as they have slightly different implementations than the other 3 module categories

shaokiat replied:
I think Sam meant that the logic behind the implementation can be abstracted out so there's not as much repeated code because now there are still repeated code in the 3 methods which is the double for loops.
I'm guessing maybe the addition of the helper function would be good.

yan-soon replied:
oh right HAHAH okay abstracted further

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+     * any undone Foundation Modules.
+     */
+    public void compareFoundation() {
+        boolean foundationClear = true;
```
Maybe we can name this `isFoundationCleared` to make it more boolean-like? Same applies for the other compareX functions

yan-soon replied:
Good call, changed it!

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+     */
+    public void compareFoundation() {
+        boolean foundationClear = true;
+        String modulesToAdd = "";
```
Perhaps we can use a java `StringBuilder` here since we're iteratively doing string concatenations (line 91). Otherwise we'll be doing O(n) operations per iteration and this fxn will run in O(n^2) :grin:

yan-soon replied:
Okay did not know about this HAHAHAH changed accordingly!

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+        }
+    }
+    @Override
+    public CommandResult execute(Model model) throws IOException, DataConversionException {
```
Maybe it'll be better if we handle the exceptions in this command instead of propagating it cos that would mean we need to change the `execute` contract (as has been done). But in doing so, we are breaking the interface segregation principle since ALL other commands are also now bound to a contract that isn't applicable to them.

yan-soon replied:
aHHHHH so this is the way....HAHAH I totally forgot I could do this. thanks man, changed accordingly :D

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+        setITprofStorage(ITPROF_PATH);
+        setMathAndScienceStorage(MATHANDSCI_PATH);
+        setScienceStorage(SCIENCE_PATH);
+        setInternshipStorage(INTERNSHIP_PATH);
```
(cont. from prev comment)
So maybe what we can is to wrap these error-throwing calls in a try-catch block and handle exceptions here instead. After all, clients expect this `execute` method to give them a result (success or failure message), and not an exception.

```suggestion
try {
        setFoundationStorage(FOUNDATION_PATH);
        setITprofStorage(ITPROF_PATH);
        setMathAndScienceStorage(MATHANDSCI_PATH);
        setScienceStorage(SCIENCE_PATH);
        setInternshipStorage(INTERNSHIP_PATH);
} catch (IOException | DataConversionException e) {
        // from the error msg you wrote in MainWindow hehe
        return new CommandResult("File not found :(");
}
```

Silvernitro commented on src/main/java/seedu/address/logic/commands/ScienceCommand.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public CommandResult execute(Model model) throws IOException, DataConversionException {
+        requireNonNull(model);
+        Path sciencePath = Paths.get("src/main/data/sciencemodules.json");
```
Maybe we can declare the pathstring as a class-level constant instead to make it obvious where the filepath is stored when we need to change it next time :D

yan-soon replied:
I changed it such that I used the static import of the SCIENCE_PATH.

Silvernitro commented on src/main/java/seedu/address/logic/commands/ScienceCommand.java:
```diff
@@ -0,0 +1,67 @@
+    public CommandResult execute(Model model) throws IOException, DataConversionException {
+        requireNonNull(model);
+        Path sciencePath = Paths.get("src/main/data/sciencemodules.json");
+        setStorage(sciencePath);
```
Same as `RequiredCommand`, perhaps we can wrap this in a try-catch and handle the exception here rather than propagating it and binding all other commands to a new contract

yan-soon replied:
Okay, changed accordingly, similar to what I did for RequiredCommand

Silvernitro commented on src/main/java/seedu/address/logic/commands/ScienceCommand.java:
```diff
@@ -0,0 +1,67 @@
+        }
+        for (Module module : modules) {
+            String moduleToAdd = module.getModuleCode() + " (" + module.getModularCredits() + " MCs)";
+            moduleNames += "\n" + moduleToAdd;
```
Same as previous comment in `RequiredCommand`, we can use a `StringBuilder` instead so that this fxn can run in O(n) instead of O(n^2)

yan-soon replied:
Rectified!

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,319 @@
+
+    private ObservableList<Module> modules;
+    private String leftOverModules = "";
+    private Optional<ReadOnlyGradPad> foundationStorage;
```
Maybe we can initialize this to an empty `Optional` instead of leaving it as null? I think it's a bit weird that the field can either be `null`, empty `Optional`, or filled `Optional` when the purpose of Optionals is to abstract away nulls anyway HAHA 

But this one is really a minor suggestion :grin:

![Screenshot 2020-10-22 at 5 05 16 PM](https://user-images.githubusercontent.com/31149701/96849889-c51c3500-1488-11eb-99f2-4435012efa48.png)

yan-soon replied:
Changed the whole class design so, no more Optional<> attributes HAHAHA

Silvernitro commented on src/test/java/seedu/address/logic/commands/RequiredCommandTest.java:
```diff
@@ -0,0 +1,162 @@
+    public void setFoundationStorageInvalidPath_returnsEmptyOptional() throws IOException, DataConversionException {
+        requiredCommand.setFoundationStorage(INVALID_PATH);
+        Optional<ReadOnlyGradPad> actual = requiredCommand.getFoundationStorage();
+        assertEquals(actual, Optional.empty());
```
Perhaps we can consider using this instead:

```suggestion
        assertTrue(actual.isEmpty());
```

As mentioned in the java API, it's generally not recommended to do equality checks for Optionals cos they produce unpredictable results 😁  Same applies for the other tests below and also `ScienceCommandTest`

yan-soon replied:
okay roger doger! changed accordingly

Silvernitro reviewed (commented):
Great job in general man, the logic is sound and the approach really makes sense, but just a few comments/suggestions :D

And also some general comments/suggestions:

- We can consider abstracting out all the storage access logic in `RequiredCommand` into a separate class like `RequiredCommandStorage`. This way, `RequiredCommand` becomes more lean since it doesn't need to store storage fields and their setters/getters. It also adheres to the single-responsibility and open-closed principle cos `RequiredCommand` doesn't need to care abt how it gets the data, and we are free to change the storage logic without touching `RequiredCommand` next time. `RequiredCommand` can then call stuff like `.getFoundationModules()`
- We can't store the JSON files in the `main/data` directory cos they won't be bundled by ShadowJar and the smoke tests won't work properly. They need to be put in the resources folder instead. BUT ITS OK I also just realised this cos the `modules.json` in Nusmods needs to move too oooops

yan-soon commented:
As per Sam's reviews, I have made some changes to the PR.

The major changes are as such:

**RequiredCommandStorage**

Added this class to deal with all the pulling of modules from the respective JSON files.

**RequiredCommandMessages**

Added this class to store all the standard messages and paths that the RequiredCommand will use.

**RequiredCommand**

- Changed the class such that it no longer has to deal with storage of the required modules! It now simply uses Getter methods from the RequiredCommandStorage to get the respective modules from the various fields.

- The class now mainly deals with all the comparison of GradPad modules against all the required modules.

**ScienceCommand**

- Changed the class slightly to improve the overall performance and design

**RequiredCommandTest**

Changed the tests to suit the new methods

**RequiredCommandStorageTest**

Added the test class to test the new `RequiredCommandStorage` class.

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,194 @@
+        for (Module module : requiredScience) {
+            if (doesModuleAlreadyExist(module, currentModules)) {
+                leftOverModules += MESSAGE_SUCCESS_SCIENCE + "\n";
+                areModulesCleared = false;
```
Shouldn't `areModulesCleared` become true when we find out that the user has taken a science module?

yan-soon replied:
accidentally flipped the booleans HAHAH changed the conditions and booleans accordingly

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,194 @@
+                areModulesCleared = false;
+                break;
+            }
+        } if (areModulesCleared) {
+            leftOverModules += MESSAGE_SCIENCE + "\n";
```
Likewise if `areModulesCleared` is true then shouldn't we be showing MESSAGE_SUCCESS_SCIENCE instead since the user has cleared this basket HAHA

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,194 @@
+            if (doesModuleAlreadyExist(module, currentModules)) {
+                int modularCredits = Integer.parseInt(module.getModularCredits().toString());
+                modularScore += modularCredits;
+                areModulesCleared = false;
```
How come taking a single internship module means that `areModulesCleared` is false ahh? I tot it should only be false if total MCs < 12 since we need >= 12 in order to clear the internship basket ><

 I FEEL LIKE IM MISSING SOMETHING OBVIOUS LOL SO PLS FORGIVE ME IF I AM HAHA IM DAMN CONFUSED

yan-soon replied:
sorry for the misleading boolean names AHAHHA changed the whole code, so no more booleans, just an If-else block to settle if the module is present or not

shaokiat reviewed (approved):
LGTM! Nice one Yan! 💯💯💯

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -0,0 +1,194 @@
+        int modularScore = 0;
+        StringBuilder leftOverInternship = new StringBuilder();
+        for (Module module : requiredInternship) {
+            boolean areModulesCleared = true;
+            if (doesModuleAlreadyExist(module, currentModules)) {
+                int modularCredits = Integer.parseInt(module.getModularCredits().toString());
+                modularScore += modularCredits;
+                areModulesCleared = false;
+            }
+            if (areModulesCleared) {
+                String moduleToAdd = module.getModuleCode() + " (" + module.getModularCredits() + " MCs)";
+                leftOverInternship.append("\n").append(moduleToAdd);
+            }
```
Actually how come we need the `areModulesCleared`flag instead of using an if-else statement ahhh? 

Just to clarify, is this code saying: if a module is taken, then add its MCs to `modularScore`. If it's not, add the module title/mc to the result to display instead.

PAISEH I'm very confused here ><

```suggestion
            if (doesModuleAlreadyExist(module, currentModules)) {
                int modularCredits = Integer.parseInt(module.getModularCredits().toString());
                modularScore += modularCredits;
            } else { 
                String moduleToAdd = module.getModuleCode() + " (" + module.getModularCredits() + " MCs)";
                leftOverInternship.append("\n").append(moduleToAdd);
            }
```

Silvernitro reviewed (commented):
Thanks for addressing the previous comments! The code looks great :D Just a few more clarifications cos I got confused at some parts ><
